### PR TITLE
Set the class of a swizzled object back to its original class upon deallocation.

### PR DESCRIPTION
**Pull request by ncooke3**
~~This just avoids disposing the swizzler's generated classes when zombies is enabled.~~

- [x] detailed explanation of what we do and why in the comments
- [ ] tests?
- [ ] document possible edge cases when it may not work as expected...

Fix [#8321](https://github.com/firebase/firebase-ios-sdk/issues/8321)

### CODE REVIEW

**Review by maksymmalyhin — COMMENTED** [review]
Looks great! Could you please the following:
- detailed explanation of what we do and why in the comments 
- tests when possible
- document possible edge cases when it may not work as expected if there are

Thanks!

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m [hunk 1]
```diff
@@ -61,4 +65,23 @@ - (BOOL)respondsToSelector:(SEL)aSelector {
   return [gulClass instancesRespondToSelector:aSelector] || [super respondsToSelector:aSelector];
 }
 
+- (void)dealloc {
+  @autoreleasepool {
+    // Set the generated class to the original class.
+    Class originalClass = [[self gul_class] superclass];
+    if (![originalClass superclass] || [self isProxy]) {
```
Comment (maksymmalyhin) [inline GoogleUtilities/ISASwizzler/GULSwizzledObject.m:72]: We will need to explain these conditions in the comments and make sure they have a good test coverage

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m [hunk 2]
```diff
@@ -61,4 +65,23 @@ - (BOOL)respondsToSelector:(SEL)aSelector {
...
+      return;
+    }
+
+    // `self` is now the original class.
+    object_setClass(self, originalClass);
```
Comment (maksymmalyhin) [inline GoogleUtilities/ISASwizzler/GULSwizzledObject.m:77]: I'm curious if zombie diagnostic still works after this.

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m (hunk 1 above)
Comment (ncooke3) [reply GoogleUtilities/ISASwizzler/GULSwizzledObject.m]: Done.

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m (hunk 2 above)
Comment (ncooke3) [reply GoogleUtilities/ISASwizzler/GULSwizzledObject.m:118]: I pulled this branch into the Performance quickstart and the quickstart did not crash (both with and without zombies enabled). Is this what you are referring to?

**Review by paulb777 — APPROVED** [review]
Great description. Thanks!

I'm fine if you want to merge it here to enable more testing.

You could update or revert if testing finds anything or @maksymmalyhin has additional comments.

**Comment by ncooke3** [comment]
Rebased after [Update tool path in test_coverage #53](https://github.com/google/GoogleUtilities/pull/53) merged

**Comment by ncooke3** [comment]
Merging and staging to integration test in nightlies

File: GoogleUtilities/ISASwizzler/GULSwizzledObject.m [hunk 3]
```diff
@@ -61,4 +65,80 @@ - (BOOL)respondsToSelector:(SEL)aSelector {
...
+    // Note: `self` is now the original class.
+
+    // If `self` is now a root class or an object inheriting from `NSProxy`, we
+    // cannot call `dealloc` on it and should return now.
+    //
+    // Additional context:
+    //
+    // A root class is a class that does not have a superclass. Foundation has
+    // two root classes: `NSObject` and `NSProxy`.
+    BOOL classIsARootClass = ![self superclass] || [self isProxy];
```
Comment (maksymmalyhin) [inline GoogleUtilities/ISASwizzler/GULSwizzledObject.m:129]: If `isProxy` is `true` for `NSProxy` subclasses? If yes, should we check if the superclass is `NSProxy` instead? Something like this should do the trick `[[self superclass] isEqual:[NSProxy class]]`, though the exact code should be validated.

**Review by maksymmalyhin — COMMENTED** [review]
Thank you for the deep investigation and great docs for the fix! I added a couple more questions I would like to be answered before releasing the change to be more comfortable with it.

Looks like existing tests already helped us to not introduce a regression. We should add tests to the new logic as well so it doesn't get accidentally broken in the next iteration of changes. 

One of the most important part we should validate is the order in which `dealloc` is called for the class hierarchy. I think we can do this by creating a a class hierarchy with classes that e.g. post a notification with their class names in dealloc. Then we can create expectations and wait for them to arrive in a specific order. Dealloc of NSObject itself we will probably have to check by side effects like nullifying weak links.

P.S. TBH, I would prefer to get rid of swizzling altogether and forget about messing with Objective-C runtime 😄
